<p style="font-size:24px"><b>Predicting Fantasy Football Stars</b></p>

I love fantasy football. In general, I love sports, but there’s something about sitting down for an entire day on Sundays and watching football with your league mates that cannot be beat. Compared to other sports, fantasy football is easily the best type of fantasy sport. Basketball and baseball have too many games in a week to constantly be updating and changing your line up, and soccer has too little actual metrics to use to determine what fantasy points are worth. However, football has the perfect mix of frequency and statistical measurement. Games aren’t on too frequently, but not too scarcely either. Points are judged by yards and touchdowns, so one play can literally make or break your week. Overall, fantasy football is very exciting and fun, and I’m hoping that I can somehow predict who I should draft next year.

<i>Cleaning Data</i>

First and foremost, I need a data set to work with. Thankfully, Funk Monarch on Kaggle posted a huge data set containing every important offensive metric for every player from every week since 2012. This was a lot of data to sort through, but ultimately, through the help of SQL queries and ChatGPT, I managed to isolate what I determined to be the important metrics for flex positions (wide receivers, running backs, and tight ends) in one spreadsheet, and quarterbacks in another spreadsheet. Now, I can begin building a predictive model.

<i>Building The Model (Flex Players)</i>

To build the actual model, I transitioned from MySQL to VS Code to use python. Python has various libraries that make building predictive models much, much easier. Also, everything in the data will be catered towards PPR scoring, so I apologize in advance to all of those in standard and half-PPR scoring leagues.

In [62]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score

First, I want to a construct a very basic base model to compare our future models against. This model will rely only on the previous year's total PPR points rankings to predict the next year's leaders. Yes, this is copy and paste, but it's what the 'experts' at ESPN do anyway.

In [63]:
def control(train_year):
    flex_data_path_control = '.../FantasyFootballPredictor/Data/oy_flex.csv'
    flex_data = pd.read_csv(flex_data_path_control)

    data_train = flex_data[flex_data['season'] == train_year]
    data_actual = flex_data[flex_data['season'] == (train_year + 1)]
    
    top_20_actual = data_actual[['name', 'fantasy_points_ppr']].sort_values(by='fantasy_points_ppr', ascending=False).head(20).reset_index(drop=True)
    top_20_actual['actual_rank'] = range(1, 21)

    control_predictions = data_train[['name', 'fantasy_points_ppr']].sort_values(by='fantasy_points_ppr', ascending=False).head(20).reset_index(drop=True)
    control_predictions['predicted_rank'] = range(1, 21)
    
    score = accuracy_test(top_20_actual, control_predictions)
    print(f"Accuracy Score (Control): {score}")
    
    return(score)

Now that that is done, we can begin to construct our own model. Obviously, for our model, using the fantasy point totals from the year before is cheating, and we will not be doing that. Instead, we will be looking to use other metrics. From the dataset provided, I ran a correlation test to see which statistical measures would be most applicable.

In [64]:
flex_data_path = '.../FantasyFootballPredictor/Data/oy_flex.csv'
flex_data = pd.read_csv(flex_data_path)
numeric_columns = flex_data.select_dtypes(include=['number'])
correlation = numeric_columns.corr()
ppr_correlation = correlation['fantasy_points_ppr'].sort_values(ascending=False).head(10)
print(ppr_correlation)

fantasy_points_ppr             1.000000
total_yards                    0.977365
ppr_ppg                        0.908767
receptions                     0.908628
receiving_yards_after_catch    0.905508
total_tds                      0.891624
targets                        0.879435
ypg                            0.873394
offense_snaps                  0.868134
receiving_yards                0.855710
Name: fantasy_points_ppr, dtype: float64


Based on the test, I decided to train my model based on total yards, receptions, receiving yards after catch, total touchdowns, targets, and PPR PPG.

In [65]:
top_features = ['total_yards','receptions','receiving_yards_after_catch','total_tds','targets','ppr_ppg']
X = numeric_columns[top_features]
y = numeric_columns['fantasy_points_ppr']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

model = LinearRegression()
model.fit(X_train, y_train)
y_pred = model.predict(X_test)

mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)
print(f"Mean Squared Error: {mse}")
print(f"R-squared: {r2}")

Mean Squared Error: 4.032655032562325
R-squared: 0.9994163208787762


<i>Testing the Algorithm</i>

Now that we have our model (which has a very high MSE but an almost perfect R^2), I want to test how strong it is. To do this, I used this model and compared it with a list of the actual top 20 players in total PPR points for that year. I then gave it an accuracy score, which awarded 2 points to the model if it got the right person in the right position that they finished, 1 point if it predicted a person to be in the top 20, but in the wrong position, and 0 if it completely missed. I then compared this with the control model's accuracy score.

In [66]:
def predict_next_year(train_year):

    data_train = flex_data[flex_data['season'] == train_year]
    data_actual = flex_data[flex_data['season'] == (train_year + 1)]
    numeric_columns_train = data_train.select_dtypes(include=['number'])

    X_train = numeric_columns_train[top_features]
    y_train = numeric_columns_train['fantasy_points_ppr']

    model = LinearRegression()
    model.fit(X_train, y_train)

    data_predict = data_train.copy()
    data_predict['predicted_fantasy_points_ppr'] = model.predict(X_train)

    top_20_predicted = []
    top_20_actual = []
    for _, row in data_predict[['name', 'predicted_fantasy_points_ppr']].sort_values(by='predicted_fantasy_points_ppr', ascending=False).iterrows():
        if row['name'] not in [player['name'] for player in top_20_predicted]:
            top_20_predicted.append({'name': row['name'], 'predicted_fantasy_points_ppr': row['predicted_fantasy_points_ppr']})
        if len(top_20_predicted) == 20:
            break

    if not data_actual.empty and 'name' in data_actual.columns:
        for _, row in data_actual[['name', 'fantasy_points_ppr']].sort_values(by='fantasy_points_ppr', ascending=False).iterrows():
            if row['name'] not in [player['name'] for player in top_20_actual]:
                top_20_actual.append({'name': row['name'], 'fantasy_points_ppr': row['fantasy_points_ppr']})
            if len(top_20_actual) == 20:
                break
    else:
        for i in range(20):
            top_20_actual.append({'name': 'NA', 'fantasy_points_ppr': 'NA'})

    top_20_predicted_df = pd.DataFrame(top_20_predicted)
    top_20_predicted_df['predicted_rank'] = range(1, 21)

    top_20_actual_df = pd.DataFrame(top_20_actual)
    top_20_actual_df['actual_rank'] = range(1, 21)

    if 'name' in top_20_predicted_df.columns and 'name' in top_20_actual_df.columns:
        if 'NA' not in top_20_actual_df['name'].values:
            score = accuracy_test(top_20_actual_df, top_20_predicted_df)
        else:
            score = 'NA'
    else:
        score = 'NA'

    combined_results = pd.DataFrame({
        'Rank': range(1, 21),
        'Projected Leaders': top_20_predicted_df['name'],
        'Projected PPR PPG': top_20_predicted_df['predicted_fantasy_points_ppr'],
        '': range(1, 21),
        'Actual Leaders': top_20_actual_df['name'],
        'Actual PPR Total': top_20_actual_df['fantasy_points_ppr']
    })

    print(f"Projected vs Actual Leaders for the {train_year + 1} Season based on {train_year} Stats:")
    print(combined_results)
    print(f"Accuracy Score: {score}")
    return score

In [67]:
def accuracy_test(actual, predicted):
    score = 0
    predicted_ranks = {row['name']: row['predicted_rank'] for index, row in predicted.iterrows()}
    actual_ranks = {row['name']: row['actual_rank'] for index, row in actual.iterrows()}
    for player in actual_ranks:
        if player in predicted_ranks:
            if predicted_ranks[player] == actual_ranks[player]:
                score += 2  # Exact rank match
            else:
                score += 1  # Player is in the top 20 but not the exact rank

    return(score)

In [68]:
def test_data():
    total_score = 0
    total_score_c = 0
    for i in range(3,13):
        score = predict_next_year(2010+i)
        total_score += score
        score_c = control(2010+i)
        total_score_c += score_c
    print(f"Total Score: {total_score}")
    print(f"Total Score (Control): {total_score_c}")
test_data()

Projected vs Actual Leaders for the 2014 Season based on 2013 Stats:
    Rank Projected Leaders  Projected PPR PPG        Actual Leaders  \
0      1    Jamaal Charles         378.354726   1     Antonio Brown   
1      2        Matt Forte         335.806182   2      Le'Veon Bell   
2      3      LeSean McCoy         327.041860   3    DeMarco Murray   
3      4  Demaryius Thomas         318.025615   4        Matt Forte   
4      5     Antonio Brown         309.136943   5  Demaryius Thomas   
5      6        A.J. Green         306.130907   6      Jordy Nelson   
6      7    Calvin Johnson         303.842195   7        Dez Bryant   
7      8      Jimmy Graham         303.591151   8    Marshawn Lynch   
8      9  Brandon Marshall         301.921998   9  Emmanuel Sanders   
9     10        Dez Bryant         294.346094  10       Julio Jones   
10    11   Knowshon Moreno         293.826109  11      Randall Cobb   
11    12    Alshon Jeffery         282.074233  12     Jeremy Maclin   
12    13

<i>Improving the Model</i>

Based on the results, my model had a combined accuracy score of 80 vs the control's score of 77. While this is only about a 4% change, it shows that using my model has improved results over simply basing it off of last year's rankings. To try and improve this model, I found a data set on Kaggle from Nick Cantalupa that has every important team metric from the last 20 years. After cleaning and fixing some inconsistent variables, I joined my 2 data sets together to get one big flex player data set. I then ran a correlation test on that to find out which team stats are most applicable.

In [69]:
flex_data_path = '.../FantasyFootballPredictor/Data/flex_onlyteamstats.csv'
flex_data = pd.read_csv(flex_data_path)
numeric_columns = flex_data.select_dtypes(include=['number'])
correlation = numeric_columns.corr()
fantasy_points_ppr_correlation = correlation['fantasy_points_ppr'].sort_values(ascending=False).head(10)
print(fantasy_points_ppr_correlation)

fantasy_points_ppr      1.000000
ppr_ppg                 0.909199
total_yards             0.133062
pass_yds                0.130776
pass_net_yds_per_att    0.125223
points                  0.124520
pass_fd                 0.122532
yds_per_play_offense    0.121345
first_down              0.121257
pass_td                 0.112466
Name: fantasy_points_ppr, dtype: float64


While the correlation is minimal (topping at only 0.13), I still want to include this data in my model as from a logistic standpoint, it makes sense to factor in how the player's team is doing to determine how well they will do.

Since the correlation is so small, I didn't stick entirely to the top correlated attributes and experimented with others. Ultimatley, I decided to add how many points that player's team scored that year ('points'), how many offensive plays they had ('plays_offense'), how many games the player played (this wasn't part of the new team data, but rather was something that came to mind while I was working on this). I decided on these because I believe that a fantasy football player is only as good as his team. If they aren't putting up yards on offense or getting near the endzone, then how can I expect the player to produce?

In [70]:
flex_data_path = '.../FantasyFootballPredictor/Data/flex_team.csv'
flex_data = pd.read_csv(flex_data_path)
numeric_columns = flex_data.select_dtypes(include=['number'])

top_features = ['ypg','total_yards','total_tds','receiving_yards_after_catch','receptions','ppr_ppg','games','points','plays_offense']
X = numeric_columns[top_features]
y = numeric_columns['fantasy_points_ppr']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

model = LinearRegression()
model.fit(X_train, y_train)
y_pred = model.predict(X_test)

mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)
print(f"Mean Squared Error: {mse}")
print(f"R-squared: {r2}")

Mean Squared Error: 2.4682105103044885
R-squared: 0.9996166784239191


Adding these metrics to my model actually lowered the MSE by 1.5, and given the size of the data set, this is very, very good. Now, I will check it against the existing data to see how it performs against the control.

In [71]:
test_data()

Projected vs Actual Leaders for the 2014 Season based on 2013 Stats:
    Rank Projected Leaders  Projected PPR PPG        Actual Leaders  \
0      1    Jamaal Charles         378.311627   1     Antonio Brown   
1      2        Matt Forte         336.666710   2      Le'Veon Bell   
2      3      LeSean McCoy         328.257916   3    DeMarco Murray   
3      4  Demaryius Thomas         318.125197   4        Matt Forte   
4      5     Antonio Brown         309.659069   5  Demaryius Thomas   
5      6        A.J. Green         306.076867   6      Jordy Nelson   
6      7    Calvin Johnson         303.644180   7        Dez Bryant   
7      8      Jimmy Graham         302.695269   8    Marshawn Lynch   
8      9  Brandon Marshall         302.091853   9  Emmanuel Sanders   
9     10   Knowshon Moreno         294.381646  10       Julio Jones   
10    11        Dez Bryant         294.346977  11      Randall Cobb   
11    12    Alshon Jeffery         283.045157  12     Jeremy Maclin   
12    13

This upgraded model performed even better than the previous one (Accuracy score of 83, 3 more than the previous 80). So far, our model has continued to grow and improve compared to the base copy-and-paste ESPN model, which is great. However, we are far from finished. Out of curiosity, I want to see if the model is skewed towards a certain position (i.e. does it favor WRs over RBs).

In [72]:
def predict_next_year_position(train_year):
    data_train = flex_data[flex_data['season'] == train_year]
    data_actual = flex_data[flex_data['season'] == (train_year + 1)]
    numeric_columns_train = data_train.select_dtypes(include=['number'])

    X_train = numeric_columns_train[top_features]
    y_train = numeric_columns_train['fantasy_points_ppr']

    model = LinearRegression()
    model.fit(X_train, y_train)

    data_predict = data_train.copy()
    data_predict['predicted_fantasy_points_ppr'] = model.predict(X_train)

    top_20_predicted = []
    top_20_actual = []
    for _, row in data_predict[['name', 'position', 'predicted_fantasy_points_ppr']].sort_values(by='predicted_fantasy_points_ppr', ascending=False).iterrows():
        if row['name'] not in [player['name'] for player in top_20_predicted]:
            top_20_predicted.append({'name': row['name'], 'position': row['position'], 'predicted_fantasy_points_ppr': row['predicted_fantasy_points_ppr']})
        if len(top_20_predicted) == 20:
            break

    if not data_actual.empty and 'name' in data_actual.columns:
        for _, row in data_actual[['name', 'position', 'fantasy_points_ppr']].sort_values(by='fantasy_points_ppr', ascending=False).iterrows():
            if row['name'] not in [player['name'] for player in top_20_actual]:
                top_20_actual.append({'name': row['name'], 'position': row['position'], 'fantasy_points_ppr': row['fantasy_points_ppr']})
            if len(top_20_actual) == 20:
                break
    else:
        for i in range(20):
            top_20_actual.append({'name': 'NA', 'position': 'NA', 'fantasy_points_ppr': 'NA'})

    top_20_predicted_df = pd.DataFrame(top_20_predicted)
    top_20_predicted_df['predicted_rank'] = range(1, 21)

    top_20_actual_df = pd.DataFrame(top_20_actual)
    top_20_actual_df['actual_rank'] = range(1, 21)

    if 'name' in top_20_predicted_df.columns and 'name' in top_20_actual_df.columns:
        if 'NA' not in top_20_actual_df['name'].values:
            score = accuracy_test(top_20_actual_df, top_20_predicted_df)
        else:
            score = 'NA'
    else:
        score = 'NA'

    combined_results = pd.DataFrame({
        'Rank': range(1, 21),
        'Projected Leaders': top_20_predicted_df['name'] + " (" + top_20_predicted_df['position'] + ")",
        'Projected PPR PPG': top_20_predicted_df['predicted_fantasy_points_ppr'],
        '': range(1, 21),
        'Actual Leaders': top_20_actual_df['name'] + " (" + top_20_actual_df['position'] + ")",
        'Actual PPR Total': top_20_actual_df['fantasy_points_ppr']
    })

    predicted_positions_count = top_20_predicted_df['position'].value_counts()
    actual_positions_count = top_20_actual_df['position'].value_counts()

    position_comparison = pd.DataFrame({
        'Position': predicted_positions_count.index.union(actual_positions_count.index),
        'Predicted Count': predicted_positions_count.reindex(predicted_positions_count.index.union(actual_positions_count.index), fill_value=0),
        'Actual Count': actual_positions_count.reindex(predicted_positions_count.index.union(actual_positions_count.index), fill_value=0)
    }).reset_index(drop=True)
    position_comparison['Difference'] = position_comparison['Predicted Count'] - position_comparison['Actual Count']

    print(f"Projected vs Actual Leaders for the {train_year + 1} Season based on {train_year} Stats:")
    print(combined_results)
    print(f"Accuracy Score: {score}")
    print("\nPosition Comparison:")
    print(position_comparison)
    
    return score, position_comparison

def test_data_position():
    total_score = 0
    total_score_c = 0
    total_position_diff = pd.DataFrame(columns=['Position', 'Predicted Count', 'Actual Count', 'Difference'])
    
    for i in range(3, 13):
        score, position_comparison = predict_next_year_position(2010 + i)
        total_score += score
        total_position_diff = pd.concat([total_position_diff, position_comparison], ignore_index=True)
        score_c = control(2010 + i)
        total_score_c += score_c
    
    total_position_diff = total_position_diff.groupby('Position').agg(
        Predicted_Count_Total=('Predicted Count', 'sum'),
        Actual_Count_Total=('Actual Count', 'sum'),
        Difference_Total=('Difference', 'sum'),
        Difference_Max=('Difference', 'max'),
        Difference_Min=('Difference', 'min')
    ).reset_index()
    
    print(f"Total Score: {total_score}")
    print(f"Total Score (Control): {total_score_c}")
    print("\nTotal Position Difference over all years:")
    print(total_position_diff)

test_data_position()


Projected vs Actual Leaders for the 2014 Season based on 2013 Stats:
    Rank      Projected Leaders  Projected PPR PPG             Actual Leaders  \
0      1    Jamaal Charles (RB)         378.311627   1     Antonio Brown (WR)   
1      2        Matt Forte (RB)         336.666710   2      Le'Veon Bell (RB)   
2      3      LeSean McCoy (RB)         328.257916   3    DeMarco Murray (RB)   
3      4  Demaryius Thomas (WR)         318.125197   4        Matt Forte (RB)   
4      5     Antonio Brown (WR)         309.659069   5  Demaryius Thomas (WR)   
5      6        A.J. Green (WR)         306.076867   6      Jordy Nelson (WR)   
6      7    Calvin Johnson (WR)         303.644180   7        Dez Bryant (WR)   
7      8      Jimmy Graham (TE)         302.695269   8    Marshawn Lynch (RB)   
8      9  Brandon Marshall (WR)         302.091853   9  Emmanuel Sanders (WR)   
9     10   Knowshon Moreno (RB)         294.381646  10       Julio Jones (WR)   
10    11        Dez Bryant (WR)         

Over all of the availible years, the model only inaccuratley predicts that 2 WRs instead of 2 RBs appear. Yes, this differs from year to year (as some years have even more inconsistent positioning), but the volatility is consistent throughout all of the positions (as shown in the max/min table).

Out of curiosity, I want to use the same approach I used in the QB Predictor to see if there's a different combination of top features that might improve my model, whose MSE currently stands around 2.5.

In [73]:
from itertools import combinations
from sklearn.feature_selection import RFE

flex_data_path = '.../FantasyFootballPredictor/Data/flex_team.csv'
flex_data = pd.read_csv(flex_data_path)
numeric_columns = flex_data.select_dtypes(include=['number'])

correlation = numeric_columns.corr()
fantasy_points_ppr_correlation = correlation['fantasy_points_ppr'].sort_values(ascending=False)

top_25_features = fantasy_points_ppr_correlation.index[1:26].tolist()  # index[0] is 'fantasy_points_ppr'

X = numeric_columns[top_25_features]
y = numeric_columns['fantasy_points_ppr']

model = LinearRegression()
rfe = RFE(model, n_features_to_select=1)
rfe.fit(X, y)

ranking = rfe.ranking_
ranked_features = pd.DataFrame({'Feature': top_25_features, 'Rank': ranking}).sort_values(by='Rank')

def evaluate_top_feature_combinations(X, y, max_features=5):
    results = []
    feature_list = ranked_features['Feature'].tolist()

    for r in range(1, max_features + 1):
        for combo in combinations(feature_list[:max_features], r):
            combo = list(combo)
            
            X_subset = X[combo]
            X_train, X_test, y_train, y_test = train_test_split(X_subset, y, test_size=0.2, random_state=42)

            model = LinearRegression()
            model.fit(X_train, y_train)
            
            y_pred = model.predict(X_test)
            mse = mean_squared_error(y_test, y_pred)
            
            results.append((combo, mse))
    
    results.sort(key=lambda x: x[1])
    
    return results

results = evaluate_top_feature_combinations(X, y, max_features=15)

print("Top 10 feature combinations based on MSE:")
for combo, mse in results[:10]:
    print(f"Features: {combo}, MSE: {mse}")


Top 10 feature combinations based on MSE:
Features: ['target_share', 'air_yards_share', 'total_tds', 'ppr_ppg', 'rushing_tds', 'receptions', 'receiving_fumbles', 'carries', 'rec_ypg', 'rush_ypg', 'touches', 'total_yards'], MSE: 1.9297512236217902
Features: ['target_share', 'air_yards_share', 'total_tds', 'ppr_ppg', 'receptions', 'receiving_fumbles', 'receiving_tds', 'carries', 'rec_ypg', 'rush_ypg', 'touches', 'total_yards'], MSE: 1.9330417760553404
Features: ['target_share', 'air_yards_share', 'total_tds', 'ppr_ppg', 'rushing_tds', 'receptions', 'receiving_fumbles', 'offense_pct', 'carries', 'rec_ypg', 'rush_ypg', 'touches', 'total_yards'], MSE: 1.9330506309118431
Features: ['target_share', 'total_tds', 'ppr_ppg', 'rushing_tds', 'receptions', 'receiving_fumbles', 'carries', 'rec_ypg', 'rush_ypg', 'touches', 'total_yards'], MSE: 1.9339479786702032
Features: ['target_share', 'air_yards_share', 'total_tds', 'ppr_ppg', 'receptions', 'receiving_fumbles', 'receiving_tds', 'offense_pct', 'ca

In [74]:
top_features =['target_share', 'air_yards_share', 'total_tds', 'ppr_ppg', 'rushing_tds', 'receptions', 'receiving_fumbles', 'carries', 'rec_ypg', 'rush_ypg', 'touches', 'total_yards']
test_data()

Projected vs Actual Leaders for the 2014 Season based on 2013 Stats:
    Rank Projected Leaders  Projected PPR PPG        Actual Leaders  \
0      1    Jamaal Charles         378.815299   1     Antonio Brown   
1      2        Matt Forte         336.677172   2      Le'Veon Bell   
2      3      LeSean McCoy         329.419619   3    DeMarco Murray   
3      4  Demaryius Thomas         318.780934   4        Matt Forte   
4      5     Antonio Brown         310.495063   5  Demaryius Thomas   
5      6        A.J. Green         306.431775   6      Jordy Nelson   
6      7    Calvin Johnson         304.580200   7        Dez Bryant   
7      8      Jimmy Graham         304.107261   8    Marshawn Lynch   
8      9  Brandon Marshall         302.650701   9  Emmanuel Sanders   
9     10   Knowshon Moreno         294.382275  10       Julio Jones   
10    11        Dez Bryant         291.644237  11      Randall Cobb   
11    12    Alshon Jeffery         281.016289  12     Jeremy Maclin   
12    13

Despite this new model having an MSE of 1.9, a 24% decrease, the accuracy score was still the same at 83.